### PR TITLE
matmul: optimize cases (nd @ 2d) to be memory and compute efficient

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. Resolves https://github.com/Lightning-AI/lightning-thunder/issues/872.

See https://github.com/Lightning-AI/lightning-thunder/issues/872#issuecomment-2283599147 for more details.

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -3364,6 +3364,20 @@ def matmul(a: TensorLike, b: TensorLike, /) -> TensorLike:
     if a.ndim == 1 or b.ndim == 1:
         return prims.matmul(a, b)
 
+    # Case nd @ 2d --> reduce to a 2d gemm
+    if a.ndim > 2 and b.ndim == 2:
+        a_batch_dims = a.shape[:-2]
+
+        # a -> a_2d by flattening batch dims with the row space
+        a_2d = a.reshape(-1, a.shape[-1])
```
Comment (t-vi) [inline thunder/torch/__init__.py:3372]: This is OK because we know that PyTorch would do it, too, right?

**Review by t-vi — APPROVED** [review]
Thank you @nikitaved

File: thunder/torch/__init__.py (hunk 1 above)
Comment (nikitaved) [reply thunder/torch/__init__.py:3372]: PyTorch will, our prim and lowering otherwise will not.

Comment (t-vi) [reply thunder/torch/__init__.py:3372]: Yeah, just that we don't introduce more copies when things are non-contiguous. Let's have it.

**Comment by mruberry** [comment]
fyi @parthmannan, @tfogal

**Pull request merged**